File: src/main/java/app/application/utils/service/YoutubePlaylistDownloadService.java
package app.application.utils.service;

import app.application.data.entities.YoutubePlaylist;
import app.application.data.entities.YoutubePlaylistVideoDetail;
import app.application.data.entities.YoutubeVideo;
import app.application.listener.YoutubePlaylistDownloadListener;
import app.application.utils.service.data.YoutubeVideoDataService;
import com.github.kiulian.downloader.downloader.request.RequestVideoFileDownload;
import javafx.application.Platform;
import javafx.scene.control.Label;
import lombok.Setter;
import lombok.SneakyThrows;
import org.springframework.stereotype.Service;

import java.io.File;
import java.nio.file.Paths;

@Service
public class YoutubePlaylistDownloadService extends YoutubeDownloadService {

    private final YoutubeVideoDataService youtubeVideoDataService;

    public YoutubePlaylistDownloadService(YoutubeVideoDataService youtubeVideoDataService) {
        this.youtubeVideoDataService = youtubeVideoDataService;
    }

    @Setter
    private Label label;

    @SneakyThrows
    public void downloadPlaylist(YoutubePlaylist youtubePlaylist){
        this.youtubeDownloadListener = new YoutubePlaylistDownloadListener(dialogManager, youtubePlaylist.getPlaylistSize());
        int size = youtubePlaylist.getPlaylistSize();
        int progress = 0;
        for (YoutubePlaylistVideoDetail video : youtubePlaylist.getPlaylistVideos()) {
            progress++;
            setLabelProgress(progress, size);
            if(!video.getIgnore().get()){
                downloadAsync(youtubePlaylist.getPlaylistTitle(), youtubeVideoDataService.getYoutubeVideo(video.getVideoId()));
            }
        }
    }

    protected void downloadAsync(String playlistTitle, YoutubeVideo youtubeVideo){
        RequestVideoFileDownload requestVideoFileDownload = new RequestVideoFileDownload(youtubeVideo.getBestVideoWithAudioFormat());
        requestVideoFileDownload.renameTo(youtubeVideo.getVideoTitle()).overwriteIfExists(true)
                .saveTo(Paths.get(userConfigHandler.getUserConfig().getDownloadDir().get() + File.separator + playlistTitle).toFile())
                .callback(youtubeDownloadListener);
        youtubeDownloader.downloadVideoFile(requestVideoFileDownload);
    }

    private void setLabelProgress(int current, int max){
        Platform.runLater(() -> label.setText("Videos: " + current + "/" + max));
    }

}

File: src/main/kotlin/app/application/spring/service/download/YoutubePlaylistDownloadService.kt
package app.application.spring.service.download

import app.application.data.entities.YoutubePlaylist
import app.application.data.entities.YoutubePlaylistVideoDetail
import app.application.data.entities.YoutubeVideo
import app.application.exception.CantDeleteFileException
import app.application.listener.YoutubePlaylistDownloadListener
import app.application.spring.service.GlobalObjectHandler
import app.application.spring.service.data.YoutubeVideoDataService
import com.github.kiulian.downloader.downloader.request.RequestVideoFileDownload
import com.github.kiulian.downloader.model.videos.formats.Format
import javafx.application.Platform
import javafx.scene.control.Label
import lombok.SneakyThrows
import org.springframework.stereotype.Service
import java.io.File
import java.nio.file.Paths
import java.util.concurrent.atomic.AtomicInteger

@Service
class YoutubePlaylistDownloadService(private val youtubeVideoDataService: YoutubeVideoDataService, private val globalObjectHandler: GlobalObjectHandler) : YoutubeDownloadService() {
    var label: Label? = null

    @SneakyThrows
    fun downloadPlaylist(youtubePlaylist: YoutubePlaylist) {
        youtubeDownloadListener = YoutubePlaylistDownloadListener(dialogManager, youtubePlaylist.playlistSize, globalObjectHandler)
        val size = youtubePlaylist.playlistSize
        val progress = AtomicInteger(0)
        youtubePlaylist.playlistVideos.stream().filter { video: YoutubePlaylistVideoDetail -> !video.ignore.get() }
                .forEach { video: YoutubePlaylistVideoDetail -> downloadVideo(youtubePlaylist, size, progress, video) }
    }

    private fun downloadVideo(
            youtubePlaylist: YoutubePlaylist,
            size: Int,
            progress: AtomicInteger,
            video: YoutubePlaylistVideoDetail) {
        progress.getAndIncrement()
        setLabelProgress(progress.get(), size)
        downloadAsync(youtubePlaylist, youtubeVideoDataService.getYoutubeVideo(video.videoId))
    }

    protected fun downloadAsync(youtubePlaylist: YoutubePlaylist, youtubeVideo: YoutubeVideo) {
        val requestVideoFileDownload = RequestVideoFileDownload(getAudioOrVideoFormat(youtubePlaylist, youtubeVideo))
        requestVideoFileDownload.renameTo(youtubeVideo.videoTitle).overwriteIfExists(true)
                .saveTo(Paths.get(
                    userConfigHandler.userConfig!!.downloadDir.get() + File.separator +
                        if (userConfigHandler.userConfig!!.subFolderForPlaylists.get()) youtubePlaylist.playlistTitle else "").toFile())
                .callback(youtubeDownloadListener)
        youtubeDownloader.downloadVideoFile(requestVideoFileDownload).data()
    }

    private fun getAudioOrVideoFormat(youtubePlaylist: YoutubePlaylist, youtubeVideo: YoutubeVideo): Format {
        return if (youtubePlaylist.audioOnly.get()) {
            youtubeVideo.audioFormat
        } else youtubeVideo.bestVideoWithAudioFormat
    }

    private fun setLabelProgress(current: Int, max: Int) {
        Platform.runLater { label!!.text = "Videos: $current/$max" }
    }
}